### PR TITLE
use fal in build.

### PR DESCRIPTION
**Pull request by vwxyzh**
@chenkennt @vicancy @superyyrrzz @qinezh @ansyral @hellosnow

### CODE REVIEW

File: src/Microsoft.DocAsCode.Build.Engine/TemplateModelTransformer.cs [hunk 1]
```diff
@@ -157,6 +157,27 @@ internal ManifestItem Transform(InternalManifestItem item)
...
+            if (EnvironmentContext.FileAbstractLayerImpl == null)
+            {
+                return null;
+            }
+            var pp = ((FileAbstractLayer)EnvironmentContext.FileAbstractLayerImpl).GetOutputPhysicalPath(fileName);
+            var expandPP = Path.GetFullPath(Environment.ExpandEnvironmentVariables(pp));
+            var outputPath = Path.GetFullPath(_context.BuildOutputFolder);
+            if (expandPP.Length > outputPath.Length &&
+                FilePathComparer.OSPlatformSensitiveStringComparer.Equals(outputPath, expandPP.Remove(outputPath.Length)) &&
+                (expandPP[outputPath.Length] == '\\' || expandPP[outputPath.Length] == '/'))
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.Engine/TemplateModelTransformer.cs]: move before equals?

File: src/Microsoft.DocAsCode.Build.Engine/TemplateModelTransformer.cs [hunk 2]
```diff
@@ -157,6 +157,27 @@ internal ManifestItem Transform(InternalManifestItem item)
...
+        {
+            if (EnvironmentContext.FileAbstractLayerImpl == null)
+            {
+                return null;
+            }
+            var pp = ((FileAbstractLayer)EnvironmentContext.FileAbstractLayerImpl).GetOutputPhysicalPath(fileName);
+            var expandPP = Path.GetFullPath(Environment.ExpandEnvironmentVariables(pp));
+            var outputPath = Path.GetFullPath(_context.BuildOutputFolder);
+            if (expandPP.Length > outputPath.Length &&
+                FilePathComparer.OSPlatformSensitiveStringComparer.Equals(outputPath, expandPP.Remove(outputPath.Length)) &&
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.Engine/TemplateModelTransformer.cs]: is it possible that outputPath uses '/' while expandPP.Remove(outputPath.Length) uses '\\'?

Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Build.Engine/TemplateModelTransformer.cs]: Path.GetFullPath will normalize it.

---
In reply to: [102364867](https://github.com/dotnet/docfx/pull/1329#discussion_r102364867) [](ancestors = 102364867)

**Comment by hellosnow** [comment]
:shipit:

**Comment by vicancy** [comment]
:shipit:

**Pull request merged**